File: SVF.py
#! /usr/bin/python

import math
import numpy as np
import sys

def victim_sim_mat(interval_a, interval_b, i, j, v_sim):
    if(i > j):
        v_sim[i][j] = DTW(interval_a, interval_b)
    else:
        v_sim[i][j] = 0

def attacker_sim_mat(interval_a, interval_b, i, j, a_sim):
    if(i > j):
        a_sim[i][j] = abs(interval_a - interval_b)
    else:
        a_sim[i][j] = 0

def DTW(interval_a, interval_b):
    dtw = np.zeros((len(interval_a), len(interval_b)))
    print interval_a
    print interval_b

    for i in range(0, len(interval_a)):
        dtw[i][0] = sys.maxsize
    for i in range(0, len(interval_b)):
        dtw[0][i] = sys.maxsize

    dtw[0][0] = 0

    for i in range(1, len(interval_a)):
        for j in range(1, len(interval_b)):
            cost = abs(interval_a[i] - interval_b[j])
            dtw[i][j] = cost + min([dtw[i-1][j], dtw[i][j-1], dtw[i-1][j-1]])

    return dtw[len(interval_a)][len(interval_b)]

def SVF(v_sim, a_sim):
    v_mean = np.mean(v_sim)
    a_mean = np.mean(a_sim)
    v_std = np.std(v_sim)
    a_std = np.std(a_sim)

    sum = 0
    for i in range(0, min(len(v_sim), len(a_sim))):
        for j in range(0, min(len(v_sim[0]), len(a_sim[0]))):
            sum += (v_sim[i][j] - v_mean)*(a_sim[i][j] - a_mean)

    sum /= (v_std * a_std)
    return sum

def main():
    v_trace = [[]]
    a_trace = []
    
    v_file = "mem_trace.out"
    a_file = "dcache_trace.out"

    with open(v_file) as f:
        for line in f:
            next = line.strip().split(" ")
            temp = []
            for item in next:
                temp.append(int(item))
            temp = filter(None, temp)
            v_trace.append(temp)
    
    with open(a_file) as f:
        for next in f:
            a_trace.append(int(next))

    print len(v_trace)
    print len(a_trace)

    v_sim = np.zeros((len(v_trace), len(v_trace)))
    a_sim = np.zeros((len(a_trace), len(a_trace)))

    for i in range(0, len(v_trace)):
        for j in range(0, len(v_trace)):
            victim_sim_mat(v_trace[i], v_trace[j], i, j, v_sim)

    for i in range(0, len(a_trace)):
        for j in range(0, len(a_trace)):
            attacker_sim_mat(a_trace[i], a_trace[j], i, j, a_sim)

    print SVF(v_sim, a_sim)

main()

File: SVF.cpp
#include <stdlib.h>
#include <stdio.h>
#include <vector>
#include <string.h>
#include "DTW.h"
#include "mat.h"

using namespace std;

/* Constants and Macros */
#define MIN(x,y) (x < y ? x : y)

/* Check arguments to the program
 * @input:  argc, the number of command line arguments to the program
 */
void check_usage(int argc, char* p)
{
    if(argc < 3)
    {
        fprintf(stderr, "Usage: %s <victim file> <attacker file>\n", p);
        exit(1);
    }
}

void init_sim(int64_t*** v, int64_t*** a, int len)
{
    *a = (int64_t**) malloc(sizeof(int64_t*) * len);
    *v = (int64_t**) malloc(sizeof(int64_t*) * len);

    for(int i = 0; i < len; i++)
    {
        (*v)[i] = (int64_t*) malloc(sizeof(int64_t) * len);
        (*a)[i] = (int64_t*) malloc(sizeof(int64_t) * len);
    }
}

// Create the victim and attacker trace vector and fill with
// values read from file
void fill_vectors(vector<vector<int64_t> > &v_trace, vector<vector<int64_t> > &a_trace, char* v_file, char* a_file)
{
    FILE* vfp = fopen(v_file, "r");
    FILE* afp = fopen(a_file, "r");
    int sentinel = 1;
    char* buff = (char*) malloc(sizeof(char) * 100000);

    // Loop through file to create trace vectors
    while(sentinel)
    {
        // Line buffer
        char* end;

        // Tokenize each line into an array to be added to v_trace
        sentinel = (fgets(buff, 100000, vfp) != NULL);
        char* next_int = strtok(buff, " ");
        vector<int64_t> v;
        while(next_int)
        {
            long l = strtoull(next_int, &end, 10);
            if(l != 0)
                v.push_back(l);
            next_int = strtok(NULL, " ");
        }
        v_trace.push_back(v);

        // Tokenize each line into an array to be added to v_trace
        sentinel &= (fgets(buff, 100000, afp) != NULL);
        next_int = strtok(buff, " ");
        vector<int64_t> a;
        while(next_int)
        {
            long l = strtoull(next_int, &end, 10);
            if(l != 0)
                a.push_back(l);
            next_int = strtok(NULL, " ");
        }
        a_trace.push_back(a);
    }
    fclose(vfp);
    fclose(afp);
    free(buff);
}

// Generate similarity matrix from intervals
void gen_sim(int64_t** v_sim, int64_t** a_sim, vector<vector<int64_t> > &v, vector<vector<int64_t> > &a)
{
    // Calculate the victim similarity based on DTW algorithm
    for(int i = 0; i < v.size(); i++)
        for(int j = 0; j < v.size(); j++)
            if(i < j)
                v_sim[i][j] = (int64_t)DTW(v[i], v[j]);
            else
                v_sim[i][j] = 0;

    // Calculate the attacker similarity based on DTW algorithm
    for(int i = 0; i < a.size(); i++)
        for(int j = 0; j < a.size(); j++)
            if(i < j)
                a_sim[i][j] = (int64_t)DTW(a[i], a[j]);
            else
                a_sim[i][j] = 0;
}

/* Calculate SVF using Pearson correlation between similarity matrices
 * @input:  v, victim similarity matrix
 *          a, attacker similarity matrix
 * @return: SVF score, in range: [-1, 1]
 */
double SVF(int64_t** &v, int v_len, int64_t** &a, int a_len)
{
    double  v_mean, a_mean,
            v_std, a_std;

    v_mean = _mean(v, v_len, v_len);
    a_mean = _mean(a, a_len, a_len);

    v_std = _std(v, v_len, v_len, v_mean);
    a_std = _std(a, a_len, a_len, a_mean);

    double sum = 0;
    int bound = MIN(v_len, a_len);

    for(int i = 0; i < bound; i++)
        for(int j = 0; j < bound; j++)
        {
            sum += (v[i][j] * a[i][j]);
            printf("v[%d][%d]: %lld\t a[%d][%d]: %lld\n", i, j, v[i][j], i, j, a[i][j]);
        }
    printf("Sum: %f\n", sum);
    printf("v_mean: %f\n", v_mean);
    printf("v_std: %f\n", v_std);
    printf("a_mean: %f\n", a_mean);
    printf("a_std: %f\n", a_std);

    sum -= (v_len * a_len * v_mean * a_mean);
    double svf = sum / (v_std * a_std);
    return svf;
}

int main(int argc, char* argv[])
{
    vector< vector<int64_t> > v_trace;
    vector< vector<int64_t> > a_trace;
    int64_t **v_sim, **a_sim;

    // Check command line arguments
    check_usage(argc, argv[0]);

    // Read from victim and attacker files
    printf("Reading input files.\n");
    fill_vectors(v_trace, a_trace, argv[1], argv[2]);

    // Initialize Similarity matrices
    init_sim(&v_sim, &a_sim, v_trace.size());

    // Generate similarity matrices by comparing interval distances
    printf("Generating similarity matrices.\n");
    gen_sim(v_sim, a_sim, v_trace, a_trace);

    // Calculate SVF from similarity matrices
    printf("Computing correlation.\n");
    double result = SVF(v_sim, v_trace.size(), a_sim, a_trace.size());

    printf("Computed SVF: %f\n", ABS(result));
}

